File: cplus/sort.cpp
#include "sort.hpp"

Sort::Sort()
{
}

Sort::~Sort()
{
}

/**
 * @brief performs bubble sort on the provided array of values. 
 * The default sorting mode is ascending.
 * Bubble sort is based on the idea of repeatedly comparing pairs of adjacent elements 
 * and then swapping their positions if they exist in the wrong order.
 * 
 * @param v 
 */
void Sort::bubblesort(vector<int> &v) const
{
  auto len = v.size();
  for (int i = 0; i < len; i++)
  {
    bool swapped = false;
   for (int j = 0; j < len - i - 1; j++)
   {
     if (v[j] > v[j + 1])
     {
       this->swap(v, j, j + 1);
       swapped = true;
     }
   }
   //After an iteration, if there is no swapping, the value of swapped will be false:=0. 
   //This means elements are already sorted and there is no need to perform further iterations.
   //This will reduce the execution time and helps to optimize the bubble sort.
   if (!swapped)
   {
     break;
   }
  }
}

/**
 * @brief  performs selection sort on the provided array of values. 
 * The default sorting mode is ascending
 * The algorithm divides the input list into two parts: 
 *  - a sorted sublist of items which is built up from left to right at the front (left) of the list and 
 *  - a sublist of the remaining unsorted items that occupy the rest of the list. 
 * 
 * @param v 
 */
void Sort::selectionsort(vector<int>&v) const
{
  int len = v.size();
  for (int i = 0; i < len; i++)
  {
    int min = i;

    for (int j = i + 1; j < len; j++)
    {
      if (v[j] < v[min] )
      {
        min = j;
      }
    }
    //only swap if new minimum value is found
    if (min != i)
    {
      this->swap(v, i, min);
    }
  }
}

/**
 * @brief performs insertion sort on the provided array of values. 
 * The default sorting mode is ascending
 * The algorithm grows the sorted array/list on each iteration
 * @param v 
 */
void Sort::insertionsort(vector<int>&v) const
{
  int len = v.size();

  for (int i = 1; i < len; i++)
  {
    int key = v[i];
    int j = i - 1;
    while (j >= 0 && v[j] > key)
    {
      //shift the larger values to the right
      v[j + 1] = v[j];
      j -= 1;
    }
    //insert the current smaller value at the right position
    v[j + 1] = key;
  }
}

/**
 * @brief performs insertion sort on the provided array of values. 
 * The default sorting mode is ascending
 * The algorithm grows the sorted array/list on each iteration
 * the recursive part replaced the outer loop of the normal insertion sort
 * 
 * @param v 
 * @param n 
 */
void Sort::recursive_insertionsort(vector<int>&v, int n) const
{
  if (n > 0)
  { 
    this->recursive_insertionsort(v, (n - 1));
    int key = v[n];
    int i = n - 1;
    while (i>=0 && v[i] > key)
    {
      //shift the larger values to the right
      v[i  + 1] = v[i];
      i -= 1;
    }
    //insert the current smaller value at the right position
    v[i + 1] = key;
  }
}

/**
 * @brief performs merge sort on the provided array of values. 
 * The default sorting mode is ascending
 * 
 * @param v 
 */
void Sort::mergesort(vector<int>&v) const
{
  int len = v.size();

  if (len > 1)
  {
    auto mid = std::next(v.begin(), (len/2));
    vector<int>left(v.begin(), mid);
    vector<int>right(mid,v.end());
    
    //sort the left sub array
    this->mergesort(left);
    //sort the right sub array
    this->mergesort(right);
    //merge the sub lists
    this->merge(v, left, right);
  }
}
/**
 * @brief merges the elements from the sublists to the final list.
 * The merging is done in Linear Time
 * 
 * @param v 
 * @param left 
 * @param right 
 */
void Sort::merge(vector<int>&v, vector<int>&left, vector<int>&right) const
{
  int size = v.size();
  int lsize = left.size();
  int rsize = right.size();
  int i = 0;
  int j = 0;

  while (i + j < size)
  {
    if (j == rsize || (i < lsize && left[i] < right[j]))
    {
      v[i + j] = left[i];
      i += 1;
    }
    else
    {
      v[i + j] = right[j];
      j += 1;
    }
  }
  
}

/**
 * @brief generic function that calls the merge sort procedure.
 *  the procedure is the resposible for performing the actual sort
 * 
 * @param v 
 */
void Sort::_mergesort_(vector<int>&v) const
{
  this->_mergesort_procedure_(v, 0, v.size() - 1);
}

/**
 * @brief performs merge sort on the provided array of values. 
 * The default sorting mode is ascending
 * 
 * @param v 
 * @param start 
 * @param end 
 */
void Sort::_mergesort_procedure_(vector<int>&v, int start, int end) const
{
  if (start < end)
  {
    int mid = (start + end)/2;
    this->_mergesort_procedure_(v, start, mid);
    this->_mergesort_procedure_(v, mid + 1, end);
    this->_merge_(v, start, mid, end);
  }
}

/**
 * @brief 
 * 
 * @param v 
 * @param start 
 * @param mid 
 * @param end 
 */
void Sort::_merge_(vector<int>&v, int start, int mid, int end) const
{
  int lsize = mid - start + 1;
  int rsize = end - mid;
  int l = 0;
  int r = 0;
  int k = start;
  vector<int>left;
  left.reserve(lsize);
  vector<int>right;
  right.reserve(rsize);

  for (int i = 0; i < lsize; i++)
  {
    left.push_back(v[i + start]);
  }
  
  for (int i = 0; i < rsize; i++)
  {
    right.push_back(v[i + mid + 1]);
  }
  
  while (l < lsize && r < rsize)
  {
    if(left[l] < right[r])
    {
      v[k] = left[l];
      l += 1;
    }
    else
    {
      v[k] = right[r];
      r += 1;
    }
    k += 1;
  }

  while (l < lsize)
  {
     v[k] = left[l];
     l += 1;
     k += 1;
  }

  while (r < rsize)
  {
     v[k] = right[r];
     r += 1;
     k += 1;
  }
}

/**
 * @brief entry function to the quick sort . It calls the quicksort procedure which is the function 
 * that perfoms the actual quick sort on the provided list/array
 * 
 * @param v 
 */
void Sort::quicksort(vector<int>&v) const
{
  this->quicksort_procedure(v, 0, v.size() - 1);
}


/**
 * @brief performs quick sort on the provided array of values. 
 * The default sorting mode is ascending
 * 
 * @param v 
 * @param start 
 * @param end 
 */
void Sort::quicksort_procedure(vector<int>&v, int start, int end) const
{
  if (start < end)
  {
    int point = this->partition(v, start, end);
    this->quicksort_procedure(v, start, point);
    this->quicksort_procedure(v, point + 1, end); 
  }
}

/**
 * @brief parttions the array such that the smallest values are always at the left side
 * of the pivot with the larger values to the right of the pivot
 * 
 * @param v 
 * @param start 
 * @param end 
 * @return int 
 */
int Sort::partition(vector<int>&v, int start, int end) const
{
  int pivot = this->median_of_three(v, start, end);
  while (true)
  {
    while (v[start] < v[pivot])
    {
      start += 1;
    }
    while (v[end] > v[pivot])
    {
      end -= 1;
    }
    
    if (start >= end)
    {
      return end;
    }
    //perform the swapp
    this->swap(v, start, end);

    start += 1;
    end -= 1;
  }

  //return a bad value if the correct partition point is not found
  //this can only happen if our implementation is wrong
  return -1;
}

/**
 * @brief 
 * 
 * @param v 
 */
void Sort::countsort(vector<int> &v) const
{
  int len = v.size();
  int max = *max_element(v.begin(), v.end());
  vector<int> aux(len);
  vector<int>count(max + 1);

  //compute the frequencies of the unique values in v
  for (int i = 0; i < len; i++)
  {
    count[v[i]] += 1;
  }

  //commulative of the frequencies -  which represents the actual position of the elements in the sorted array/list
  for (int i = 1; i <= max; i++)
  {
    count[i] += count[i - 1]; 
  }

  // build sorted array
  for (int i = len - 1; i>=0; i--)
  {
    aux[--count[v[i]]] = v[i];
  }
  
  //copy back the sorted array
  v = aux;
}

/**
 * @brief performs radix sort on the provided array of values.
 * The default sorting mode is ascending
 *
 * @param v
 */
void Sort::radixsort(vector<int>&v) const
{
  int len = v.size();
  int max = *max_element(v.begin(), v.end());
  int digits = this->getdigits(max);
  int pos = 1;

  for (int i = 0; i < digits; i++)
  {
    this->radix_countsort(v, len, pos);
    pos*=10;
  }
}

/**
 * @brief performs the actual sort which is consumed by the radix sort
 * 
 * @param v 
 * @param len 
 * @param pos 
 */
void Sort::radix_countsort(vector<int>&v, int len, int pos) const
{
  vector<int>aux(len);
  vector<int>count(10);
  
  //the frequency counts
  for (int i = 0; i < len; i++)
  {
    count[((v[i] / pos) % 10)] += 1;
  }
  
  //commulative counts
  for (int i = 1; i < 10; i++)
  {
    count[i] += count[i -1];
  }
  
  //build sorted array
  for (int i = len - 1; i >= 0; i--)
  {
    aux[--count[((v[i] / pos) % 10)]] = v[i];
  }

  // copy back the sorted array
  v = aux;
}

/**
 * @brief performs heap sort on the provided array of values.
 * The default sorting mode is ascending
 *
 * @param v
 */
void Sort::heapsort(vector<int> &v) const
{
  //build the maxheap
  int len = v.size();
  this->buildmaxheap(v, len);

  //call the heapify function after swapping the values to maintain the max-heap properties
  for (int i = len - 1; i >= 0; i--)
  {
    this->swap(v, 0, i);
    this->maxheapify(v, i - 1, 0);
  }
}

/**
 * @brief corrects the heap property if any violation occurs
 * 
 * @param v 
 * @param len 
 * @param i 
 */
void Sort::maxheapify(vector<int> &v, int len, int i) const
{
  int j = i;
  int l = 2 * i;
  int r = (2 * i) + 1;

  if (l <= len && v[l] > v[j])
  {
    j = l;
  }
  if (r <= len && v[r] > v[j])
  {
    j = r;
  }

  if (i != j)
  {
    this->swap(v, i, j);
    this->maxheapify(v, len, j);
  }
}

/**
 * @brief builds the max heap that satisfies the a maxheap data structure
 *
 * @param v
 */
void Sort::buildmaxheap(vector<int> &v, int len) const
{
  for (int i = (len / 2); i >= 0; i--)
  {
    this->maxheapify(v, len - 1, i);
  }
}


/**
 * @brief compute the number of digits that constitutes the list
 * 
 * @param max 
 * @return int 
 */
int Sort::getdigits(int max) const
{
  int digits = 0;
  while (max !=0)
  {
    max /= 10;
    digits += 1;
  }
  return digits;
}

/**
 * @brief computes the median of the three values
 * - which are presented by the parameters: start, end, and (start+end)/2
 *
 * @param v
 * @param start
 * @param end
 * @return int
 */
int Sort::median_of_three(vector<int> &v, int start, int end) const
{
  int mid = (start + end)/2;
  int x = v[mid] - v[start];
  int y = v[start] - v[end];
  int z = v[mid] - v[end];

  return (x*y > 0 ? start: (x*z > 0 ? end : mid));
}

/**
 * @brief performs swap on the list by swapping the elements at the specified indices - i and j
 * 
 * @param list 
 * @param a 
 * @param b 
 */
void Sort::swap(vector<int> &v, int i, int j) const
{
  int temp = v[i];
  v[i] = v[j];
  v[j] = temp;
}

File: cshap/Sort.cs

namespace SoringAlgorithms;

public class Sort
{
  public Sort()
  {}

 /// <summary>
 /// performs bubble sort on the provided array of values.The default sorting mode is ascending.
 /// Bubble sort is based on the idea of repeatedly comparing pairs of adjacent elements
 /// and then swapping their positions if they exist in the wrong order.
 /// </summary>
 /// <param name="arr"></param>
  public void BubbleSort(int[] arr)
  {
    for (int i = 0; i < arr.Length; i++)
    {
      bool isSwapped = false; 
      for (int j = 0; j < arr.Length - i - 1; j++)
       {
         if (arr[j] > arr[j + 1])
         {
           this.Swap(arr, j, j+1);
           isSwapped = true;
         }
      }
      //exit the loop if no swap was made. Indicating that the array is already sorted
      if(!isSwapped)
      {
       break;
      }
    }
  }

  /// <summary>
  /// performs selection sort on the provided array of values. 
  ///  The default sorting mode is ascending
  /// The algorithm divides the input list into two parts: 
  ///  - a sorted sublist of items which is built up from left to right at the front(left) of the list and 
  /// - a sublist of the remaining unsorted items that occupy the rest of the list.
  /// </summary>
  /// <param name="arr"></param>
  public void SelectionSort(int[] arr)
  {
    for (int i = 0; i < arr.Length; i++)
    {
      int min = i;
      for (int j = (i + 1) ; j <arr.Length; j++)
      {
        if(arr[j] < arr[min])
        {
          min = j;
        }
      }
      //swap the vaules if the current value is the minimum
      if (i != min)
      {
        this.Swap(arr, i, min);
      }
    }
  }

  /// <summary>
  /// performs insertion sort on the provided array of values. 
  /// The default sorting mode is ascending
  // The algorithm grows the sorted array/list on each iteration
  /// </summary>
  /// <param name="arr"></param>
  public void InsertionSort(int[] arr)
  {
   for (int i = 1; i < arr.Length; i++)
   {
     int key = arr[i];
     int j = i - 1;

     while (j >=0 && arr[j] > key)
     {
        //shift the larger values to the right
        arr[j + 1] = arr[j];
       j -= 1;
     }

     //insert the new minimum at the right position
     arr[j + 1] = key;
   }
  }

  /// <summary>
  /// performs insertion sort on the provided array of values. 
  /// The default sorting mode is ascending
  /// The algorithm grows the sorted array/list on each iteration
  /// 
  /// The recursive part replaced the outer loop of the normal insertion sort
  /// </summary>
  /// <param name="arr"></param>
  /// <param name="i"></param>
  public void RecursiveInsertionSort(int[] arr, int i)
  {
    if (i > 0)
    {
      this.RecursiveInsertionSort(arr, i - 1);
      int key = arr[i];
      int j = i - 1;

      while (j >= 0 && arr[j] > key)
      {
        arr[j + 1] = arr[j];
        j -= 1;
      }
      arr[j + 1] = key;
    }
  }

  /// <summary>
  /// performs merge sort on the provided array of values. 
  /// The default sorting mode is ascending
  /// </summary>
  /// <param name="arr"></param>
  public void MergeSort(int[] arr)
 {
   int Length = arr.Length;
   if (Length > 1)
   {
    int Mid = Length/2;
    int[] left = arr.Take(Mid).ToArray();
    int[] right = arr.Skip(Mid).ToArray();

    this.MergeSort(left);
    this.MergeSort(right);
    this.Merge(arr, Length, left, left.Length,  right, right.Length);
   }
  }

  /// <summary>
  ///  merges the sublist back into the sorted list
  /// </summary>
  /// <param name="arr">array to be sorted</param>
  /// <param name="Length">Size of the original array</param>
  /// <param name="left">left sub array</param>
  /// <param name="L"> Size of the left sub array</param>
  /// <param name="right">right sub array</param>
  /// <param name="R">Size of the Right sub array</param>
  private void Merge(int[] arr, int Length, int[] left, int L, int[] right, int R)
  {
    int i = 0;
    int j = 0;

    while(i + j < Length)
    {
      if (j == R || (i < L && left[i] < right[j]))
      {
       arr[i + j] = left[i];
       i += 1;
      }
      else
      {
        arr[i + j] = right[j];
        j += 1;
      }
    }
  }

 /// <summary>
 /// performs merge sort on the provided array of values. 
 /// The default sorting mode is ascending
 /// </summary>
 /// <param name="arr"></param>
 /// <param name="start"></param>
 /// <param name="end"></param>  
 public void MergeSort(int[] arr, int start, int end)
  {
    if(start < end)
    {
      int mid = (start + end) / 2;
      this.MergeSort(arr, start, mid);
      this.MergeSort(arr, mid + 1, end);
      this.Merge(arr, start, mid, end);
    }
 }

 /// <summary>
 /// performs merge sort on the provided array of values. 
 /// The default sorting mode is ascending
 /// </summary>
 /// <param name="arr"></param>
 /// <param name="start"></param>
 /// <param name="mid"></param>
 /// <param name="end"></param>
 private void Merge(int[] arr, int start, int mid, int end)
 {
   int L = (mid - start) + 1;
   int R = end - mid;
   int l = 0;
   int r = 0;
   int k = start;

   //sub arrays
   int[] left = new int[L];
   int[] right = new int[R];

   //fill the sub arrays
   for (int i = 0; i < L; i++)
   {
     left[i] = arr[start + i];
   }
   for (int i = 0; i < R; i++)
   {
     right[i] = arr[mid + i + 1];
   }

   //sort the arrays
   while (l < L && r < R)
   {
     if (left[l] < right[r])
     {
       arr[k] = left[l];
       l += 1;
     }
     else
     {
       arr[k] = right[r];
       r += 1;
     }
     k += 1;
   }

   while (l < L)
   {
     arr[k] = left[l];
     l += 1;
     k += 1;
   }
   while (r < R)
   {
     arr[k] = right[r];
     r += 1;
     k += 1;
   }
 }

/// <summary>
/// entry function to the quick sort . It calls the quicksort procedure which is the function 
/// that perfoms the actual quick sort on the provided list/array
/// </summary>
/// <param name="arr"></param>
 public void QuickSort(int[] arr)
 {
   this.QuickSortProcedure(arr, 0, arr.Length - 1);
 }

 /// <summary>
 /// performs quick sort on the provided array of values. 
 /// The default sorting mode is ascending
 /// </summary>
 /// <param name="arr"></param>
 /// <param name="start"></param>
 /// <param name="end"></param>
 private void QuickSortProcedure(int[] arr, int start, int end)
 {
   if(start < end)
   {
     int point = this.Partition(arr, start, end);
     this.QuickSortProcedure(arr, start, point);
     this.QuickSortProcedure(arr, point + 1, end);
   }
 }

 /// <summary>
 /// parttions the array such that the smallest values are always at the left side
 /// of the pivot with the larger values to the right of the pivot
 /// </summary>
 /// <param name="arr"></param>
 /// <param name="start"></param>
 /// <param name="end"></param>
 /// <returns></returns>
 private int Partition(int[] arr, int start, int end)
 {
   int pivot = this.ThreeMedian(arr, start, end);
   while(true)
   {
     while(arr[start] < arr[pivot])
     {
       start += 1;
     }
     while (arr[end] > arr[pivot]) 
     {
       end -= 1;
     }
     if (start >= end) 
     {
       return end;
     }
     Swap(arr, start, end);
     start += 1;
     end -= 1;
   }
 }
 
 /// <summary>
 /// 
 /// </summary>
 /// <param name="arr"></param>
 public void CountSort(int[] arr)
 {
   int max = arr.Max();
   int Length = arr.Length;
   //holds the values of the sorted array
   int[] aux = new int[Length];
   //holds the frequency of the unque values
   int[] count = new int[max + 1];

   //compute the frequencies
   foreach (var item in arr)
   {
     count[item]+=1;
   }

   //cummulative of the frquencies - represents the final location of the elements 
   //in the sorted array
   for (int i = 1; i <= max; i++)
   {
     count[i] += count[i - 1];
   }

  //build final sorted array
  for (int i = Length - 1; i >= 0; i--)
  {
   aux[--count[arr[i]]] = arr[i];
  }
  
   //copy back sorted array
   Array.Copy(aux,arr,Length);
 }

 /// <summary>
 /// performs radix sort on the provided array of values.
 /// The default sorting mode is ascending
 /// </summary>
 /// <param name="arr"></param>
 public void RadixSort(int[] arr)
 {
  //largest value in the array
  int max = arr.Max();
  //number of digits making the largest value
  int digits = this.GetDigits(max);
  int Length = arr.Length;
  //tracks the current position of the values being compared
  int pos = 1;

  //perform the sort by calling the count sort 
  for (int i = 0; i < digits; i++)
  {
    this.RadixCountSort(arr, Length, pos);
    pos *= 10;
  }
 }

 /// <summary>
 /// performs the actual sort which is consumed by the radix sort
 /// </summary>
 /// <param name="arr"></param>
 /// <param name="Length"></param>
 /// <param name="pos"></param>
 private void RadixCountSort(int[] arr, int Length, int pos)
 {
  //holds the values of the sorted array
  int[] aux = new int[Length];
  //holds the frequency of the unque values
  int[] count = new int[10];

  //compute the frequencies
  foreach (var item in arr)
  {
    count[((item/pos)%10)]+=1;
  }

  //cummulative of the frquencies - represents the final location of the elements 
  //in the sorted array
  for (int i = 1; i < 10; i++)
  {
    count[i] += count[i - 1]; 
  }

  //build final sorted array
  for (int i = Length - 1; i >= 0 ; i--)
  {
   aux[--count[((arr[i] / pos) % 10)]] = arr[i];
  }

  //copy back sorted array
  Array.Copy(aux, arr, Length);
 }


  /// <summary>
  /// 
  /// </summary>
  /// <param name="arr"></param>
  public void HeapSort(int[] arr)
  {
    int Length = arr.Length;
    //build the max-heap
    this.BuildMaxHeap(arr, Length);
    //sort
    for (int i = Length - 1; i >= 0 ; i--)
    {
      this.Swap(arr,0,i);
      this.MaxHeapify(arr, i - 1, 0);
    }
  }

  /// <summary>
  /// Builds a heap that conforms to a max-heap
  /// </summary>
  /// <param name="arr"></param>
  /// <param name="length"></param>
 private void BuildMaxHeap(int[] arr, int length)
 {
  for (int i = (length / 2); i >= 0 ; i--)
  {
    this.MaxHeapify(arr, length - 1, i);      
  }
 }

 /// <summary>
 /// corrects the heap property if any violation occurs
 /// This procedure makes use of the selection sort
 /// </summary>
 /// <param name="arr"></param>
 /// <param name="Length"></param>
 /// <param name="i"></param>
 private void MaxHeapify(int[] arr, int Length, int i)
 {
  //current largest node element
  int largest = i;
  //left child node
  int left = i * 2;
  //right child node
  int right = i * 2 + 1;

  if (left <= Length && arr[left] > arr[largest])
  {
   largest = left;
  }
  if (right <= Length && arr[right] > arr[largest])
  {
   largest = right;
  }
  
  if (i != largest)
  {
    this.Swap(arr, i, largest);
    this.MaxHeapify(arr, Length, largest);
  }
 }

 /// <summary>
 /// Computes the number of digits that makes up the provided param value
 /// </summary>
 /// <param name="max"></param>
 /// <returns></returns>
 private int GetDigits(int max)
 {
  int digits = 0;
  while (max != 0)
  {
      max /= 10;
    digits += 1;
  }
  
  return digits;
 }

 /// <summary>
 /// computes the median of the three values
 /// - which are presented by the parameters: start, end, and(start+end)/2
 /// </summary>
 /// <param name="arr"></param>
 /// <param name="start"></param>
 /// <param name="end"></param>
 /// <returns></returns>
 private int ThreeMedian(int[] arr, int start, int end)
 {
   int mid = (start + end)/2;
   int x = arr[mid] - arr[start];
   int y = arr[start] - arr[end];
   int z = arr[mid] - arr[end];

   return (x * y > 0 ? start : (x * z > 0 ? end : mid));
 } 

 /// <summary>
 /// swaps the arr values specified by the param indices
 /// </summary>
 /// <param name="arr"></param>
 /// <param name="i"></param>
 /// <param name="j"></param>
 private void Swap(int[] arr, int i, int j)
  {
    int temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
  }
}